File: frontend/src/types/report.ts
export interface ReportFilterSpec {
  key: string
  label: string
  type: 'master' | 'string' | 'date'
  master_key?: string | null
  required: boolean
}

export interface ReportMeta {
  slug: string
  name: string
  description: string
  icon: string
  filters: ReportFilterSpec[]
}

export interface ReportColumn {
  key: string
  label: string
  format?: string | null
}

export interface SubSheet {
  columns: ReportColumn[]
  rows: Record<string, unknown>[]
  summary?: Record<string, unknown>
}

export interface ReportRunResult {
  slug: string
  name: string
  filters_applied: Record<string, unknown>
  columns: ReportColumn[]
  rows: Record<string, unknown>[]
  summary: Record<string, unknown>
  sheets: Record<string, SubSheet>
  generated_at: string
}

export interface ReportRunHistoryEntry {
  id: number
  timestamp: string
  report_slug: string
  report_name: string
  format: string
  row_count: number
  filters: Record<string, unknown>
  user_id: number | null
  user_email: string | null
  user_role: string | null
  user_name: string | null
}

export interface ReportHistoryList {
  items: ReportRunHistoryEntry[]
  total: number
}

export type ReportFormat = 'xlsx' | 'csv'

File: backend/app/schemas/report.py
from datetime import datetime
from typing import Any
from pydantic import BaseModel


class ReportFilterSpec(BaseModel):
    key: str
    label: str
    type: str  # "master" | "string" | "date"
    master_key: str | None = None
    required: bool = False


class ReportMeta(BaseModel):
    slug: str
    name: str
    description: str
    icon: str
    filters: list[ReportFilterSpec]


class ReportColumn(BaseModel):
    key: str
    label: str
    format: str | None = None


class ReportRunResult(BaseModel):
    slug: str
    name: str
    filters_applied: dict[str, Any]
    columns: list[ReportColumn]
    rows: list[dict[str, Any]]
    summary: dict[str, Any]
    sheets: dict[str, dict[str, Any]] = {}  # name -> {columns, rows}
    generated_at: datetime


class ReportRunOut(BaseModel):
    id: int
    timestamp: datetime
    report_slug: str
    report_name: str
    format: str
    row_count: int
    filters: dict[str, Any]
    user_id: int | None
    user_email: str | None
    user_role: str | None
    user_name: str | None

    class Config:
        from_attributes = True


class ReportHistoryList(BaseModel):
    items: list[ReportRunOut]
    total: int
